docs/data/migration/migration-data-grid-v6/migration-data-grid-v6.md

```diff
@@ -90,9 +90,24 @@ Below are described the steps you need to make to migrate from v6 to v7.
 
 - -->
 
-<!-- ### Columns
+### Columns
 
-- -->
+- The `GridColDef['type']` has been narrowed down to only accept the built-in column types.
+  TypeScript users will need to make sure to use the `GridColDef` interface when defining columns:
+
+  ```tsx
+  // 🛑 `type` is casted to `string` which is too wide
+  const columns = [{ type: 'number', id: 'field' }];
+  <DataGrid columns={columns} />;
```

Comment:
I don't like the error here though:

<img width="1047" src="https://github.com/mui/mui-x/assets/13808724/6499f812-5091-433c-b105-569decaf4a17">

I thought I could improve this by making `GridColDef` a union discriminated by the `type` field, but `type` is optional and it doesn't seem to work as expected from discriminated union :/

Reply:
If you want autocomplete without this problem, you can use `field: GridColType | string` instead of `field: GridColType`.
But of course you are loosing the actual type checking.

Otherwise, people need to set `type: 'number' as const` or `columns: GridColDef`.

Reply:
I just realized that the code snippet above was wrong.
What I wanted to show here is this error:

<img width="1008" src="https://github.com/mui/mui-x/assets/13808724/430de8e1-5a8a-44a9-a9b8-aa4b5bbf1641">


> Otherwise, people need to set type: 'number' as const or columns: GridColDef.

Yes, this is fine, my concern is the error when the `type` is inferred as `string`.
I find this part confusing:

>Types of property 'type' are incompatible.
>       Type 'string' is not assignable to type '"singleSelect"'.

Ideally, it should be:

>Types of property 'type' are incompatible.
>       Type 'string' is not assignable to type "string" | "number" | "date" | "dateTime" | "boolean" | "singleSelect" | "actions".

Reply:
Oh right, 
AFAIK, it's due to the structure of your type
You are doing:
`GridColDef = { type: 'number', ...rest } | { type: 'string', ...rest }`
To have the correct autocompletion you would need to have something like:
`GridColDef<T extends 'number' | 'string'> = { type: T' } & T extends 'string' ? { ...restString } : { ...restNumber }`
But I suppose it's not a trivial change.

Other than that, they might be some hack you could do, but I don't know them :p

Reply:
Yeah, I guess this is just the nature of unions in TS.
When `columns` is not explicitly typed, TS requires it to be compatible with all the subtypes of the `GridColDef` union.
Here's a minimal TS playground reproducing the issue: https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgOJWAEwEJwM4QDCA9gDYAiEMyA3gFDKPJgCeADhAFzIDkeYGEAHMeAbgZMYwCKUzd+goeIC+dOqEixEKdFgCCCMMGIg8JClVoTGrDtx6IjJvGOvIpMucgWglboRBgBk6m3AAUAJTIALwAfMhwICwA2gC6Kmq2OhiY5pTU0Wg5uAR5lgA+RfqGxqZlMOJ0mBAIpHBQKAjOYO4g4V2kAK4AtqFVuWT5UXEJSY1dpj0DI6YxtMzsXLw+wjwANO7Ssvb4mKc8yMriMCBhy6N4EUA